### PR TITLE
[HttpFoundation] Add a way to anonymize IPs

### PR DESCRIPTION
**Pull request by Seldaek**
| Q             | A
| ------------- | ---
| Branch?       | 4.4 
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        | symfony/symfony-docs#...  TODO

This is helpful for GDPR compliance reasons, and it isn't much code saved but it's also good if you don't have to think about how to do it.

### CODE REVIEW

File: src/Symfony/Component/HttpFoundation/Tests/IpUtilsTest.php [hunk 1]
```diff
@@ -101,4 +101,25 @@ public function invalidIpAddressData()
...
+    }
+
+    public function anonymizedIpData()
+    {
+        return [
+            ['192.168.1.1', '192.168.1.0'],
+            ['1.2.3.4', '1.2.3.0'],
+            ['2a01:198:603:0:396e:4789:8e99:890f', '2a01:198:603:0:396e:4789:8e99:0'],
+            ['0:0:0:0:0:0:0:1', '0:0:0:0:0:0:0:0'],
+            ['0:0:603:0:396e:4789:8e99:0001', '0:0:603:0:396e:4789:8e99:0'],
```
Comment (Seldaek) [inline src/Symfony/Component/HttpFoundation/Tests/IpUtilsTest.php]: For IPv6, maybe we should remove more bits, I am not sure what would be considered anonymous enough. If someone feels like researching what would be appropriate be my guest :)

File: src/Symfony/Component/HttpFoundation/IpUtils.php [hunk 2]
```diff
@@ -153,4 +153,12 @@ public static function checkIp6($requestIp, $ip)
 
         return self::$checkedIps[$cacheKey] = true;
     }
+
+    /**
+     * Anonymizes an IP/IPv6 by removing its last bit
```
Comment (xabbuh) [inline src/Symfony/Component/HttpFoundation/IpUtils.php]: Shouldn't "bit" be "byte" here?

Comment (Seldaek) [reply src/Symfony/Component/HttpFoundation/IpUtils.php]: Maybe, would like to first figure out what to do with IPv6s

Comment (javiereguiluz) [reply src/Symfony/Component/HttpFoundation/IpUtils.php]: Maybe we could use an alternative anonymizing method based on IP masks:

```php
public static function anonymize(string $ip): string
{
    $packedAddress = inet_pton($ip);
    $anonymizingMask = 4 === strlen($packedAddress) ? '255.255.255.0' : 'ffff:ffff:ffff:ffff:0000:0000:0000:0000';

    return inet_ntop($packedAddress & inet_pton($anonymizingMask));
}
```

Code adapted from https://github.com/geertw/php-ip-anonymizer/blob/master/src/IpAnonymizer.php (MIT license)

Comment (stof) [reply src/Symfony/Component/HttpFoundation/IpUtils.php]: this comment is still true today. Even for IPv4, it removes a byte, not a bit

File: src/Symfony/Component/HttpFoundation/Tests/IpUtilsTest.php [hunk 3]
```diff
@@ -101,4 +101,25 @@ public function invalidIpAddressData()
...
+    {
+        $this->assertSame($expected, IpUtils::anonymize($ip));
+    }
+
+    public function anonymizedIpData()
+    {
+        return [
+            ['192.168.1.1', '192.168.1.0'],
+            ['1.2.3.4', '1.2.3.0'],
+            ['2a01:198:603:0:396e:4789:8e99:890f', '2a01:198:603:0:396e:4789:8e99:0'],
```
Comment (stof) [inline src/Symfony/Component/HttpFoundation/Tests/IpUtilsTest.php]: more cases should be tested, especially regarding ipv6. See https://github.com/geertw/php-ip-anonymizer/pull/3 for some cases used there.

File: src/Symfony/Component/HttpFoundation/IpUtils.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpFoundation/IpUtils.php]: I think so :)

**Comment by Seldaek** [comment]
Ok I updated wording/docblock, added tests and support for IPv4-mapped IPv6s (which complicates things a little, if anyone sees a better way to do this feel free, as long as the tests pass..), and now remove the last 8 bytes (/64 equivalent) of IPv6s.

For me it's ready to ship now but any feedback is appreciated.

**Review by nicolas-grekas — APPROVED** [review]

**Review by chalasr — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @Seldaek.

**Pull request merged**